### PR TITLE
GCS_MAVLink: improved battery cell monitoring

### PR DESCRIPTION
**Pull request by tridge**
this fixes two issues:
- we are not reporting the sag corrected voltage to GCS when we are sending individual cells
- we don't cope with having more than AP_BATT_MONITOR_CELLS_MAX cells (or 12 for low flash boards, 14 for 2M boards)

it fixes this by distributing the extra voltage over the cells.

This change is particularly important for high cell count DroneCAN smart batteries, where we currently would not handle more than 14 cells and the GCS would display the wrong voltage. Even if the battery itself reports more than 14 cells (max in DroneCAN is 255 cells) then we discard the cells over AP_BATT_MONITOR_CELLS_MAX in the driver.

the PR also cleans up the use of volts vs mVolts for the local variables

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -241,36 +241,74 @@ void GCS_MAVLINK::send_battery_status(const uint8_t instance) const
...
+        /*
+          now adjust voltages to cope with two things:
+             1) we may be reporting sag corrected voltage
+             2) the battery may have more cells than can be reported by the backend, so the actual voltage may be higher than the sum
+        */
+        const float voltage_mV = battery.gcs_voltage(instance) * 1e3f;
+        float voltage_mV_sum = 0;
+        uint8_t non_zero_cell_count = 0;
+        for (uint8_t i=0; i<MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN; i++) {
+            if (cell_mvolts[i] > 0) {
```
Comment (WickedShell) [inline libraries/GCS_MAVLink/GCS_Common.cpp:271]: These will almost never be 0. An unknown cell is stored as `0xFFFF`.

**Review by WickedShell — CHANGES_REQUESTED** [review]
It looks like there's confusion on the values stored for unknown cells in the array.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 1 above)
Comment (tridge) [reply libraries/GCS_MAVLink/GCS_Common.cpp:271]: right, almost never zero, but it can be zero, and I wanted to handle that

Comment (WickedShell) [reply libraries/GCS_MAVLink/GCS_Common.cpp:271]: Sorry, I should be more clear on the concern I have. A cell that isn't monitored will be `0xFFFF` so on a 4S pack like Solo for example the last 8 cells will all be non 0, and you will accumulate a ton of extra voltage that isn't actually in the system. I believe this causes an error when you try to deal with extra voltage that may need to be distributed below.

Comment (tridge) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: ok, I see, I had thought the 65535 magic was only for mavlink, I now see SMBUS also does it, so I've changed this code to not include any 65535 values in the sum

**Review by hendjoshsr71 — COMMENTED** [review]
One thing I note that might be interesting in future versions is to distribute the extra voltage only at the end cells first moving up. (so not evenly) Then we would have the actual cell voltages for a few of the cells at least. 

Or add an extension to know what the additional amount is?

However, I imagine this year the new battery mavlink messages will get in  upstream. Those will properly split cell voltages to their own message with many instances.

**Review by peterbarker — APPROVED** [review]

**Pull request merged**